### ISSUE TITLE
checkbox and radio button not displayed correctly in screenshots

### ISSUE THREAD

**Issue by matt-riemer** [thoughtbot/capybara-webkit#871]
Hey all,

Firstly, thank you for this project!  capybara-webkit is an excellent gem, a huge benefit to the community and has become an irreplaceable tool for this developer.

I have one "small" issue that I'm truly hoping would be a quick fix for someone familiar with this codebase:

checkboxes and radio buttons are not displayed correctly in screenshots.

When using `page.save_screenshot('something.png')` the pages all render correctly, except for checkboxes and radio buttons are displayed in the top-left corner.

![screenshot_2016-01-06-13-18-19_144_png__png_image__1024 x 768_pixels_](https://cloud.githubusercontent.com/assets/954468/12154063/92cb8ac4-b47a-11e5-8ae1-f63a2762bf01.png)

I'm running:

capybara-webkit 1.7.1
capybara 2.4.1
minitest 5.8.3
minitest-capybara 0.7.2
minitest-rails-capybara 2.1.1

I compiled capybara-webkit on  Mac OS X Yosemite  10.10.5

gcc -v gives me:
```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer//usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```

Any help or pointers for a fix would be much appreciated

Thank you capybara-webkit team!

**Comment by twalpole** [comment]
What version of Qt did you compile capybara-WebKit with, and what css is positioning the checkboxes?

**Comment by jferris** [comment]
I'm glad to hear you're finding capybara-webkit so useful.

A few questions:

* What versions of Qt/WebKit are you running on? (You can see these by running `puts page.driver.version`.)
* Can you give an example of what your markup for that form looks like?
* Do you have custom CSS on `input` elements that could affect positioning?
* If you use the `selenium` driver, do the checkboxes and radios render as you'd expect?

I tried this out with a barebones form, and the elements are positioned like I'd expect.

``` html
<html>
  <body>
    <form>
      <h1>Test Form</h1>
      <div>
        <label>
          <input type="checkbox" id="input1">
          Checkbox inside label
        </label>
      </div>
      <div>
        <input type="checkbox" id="input2">
        <label for="input2">Checkbox outside label</label>
      </div>
      <div>
        <input type="radio" name="group1" id="input3">
        <label for="input3">
          Radio outside label
        </label>
        <input type="radio" name="group1" id="input4">
        <label for="input4">Radio outside label</label>
      </div>
      <div>
        <label>
          <input type="radio" name="group2" id="input5">
          Radio inside label
        </label>
        <label>
          <input type="radio" name="group2" id="input6">
          Radio inside label
        </label>
      </div>
    </form>
  </body>
</html>
```

Here's what it looks like for me:

![screenshot](https://cloud.githubusercontent.com/assets/7591/12154495/d3f8bffe-b48d-11e5-8033-48b466732c37.png)

**Comment by matt-riemer** [comment]
Thanks for the quick reply!

It's good to see that checkboxes/radio buttons are being displayed correctly on your machine.

I've run into this issue on multiple rails projects, so it's probably a machine-specific configuration issue.

Running page.driver.version gives me:
```
capybara: 2.4.1
capybara-webkit: 1.7.1
Qt: 5.4.1
WebKit: 538.1
QtWebKit: 5.4.1
````

I copied and pasted your form html (thank you!) onto an empty page with zero css.

It renders correctly for me when I run with the selenium-webdriver:

![testing_checkboxes_png_ _lightdeskapp](https://cloud.githubusercontent.com/assets/954468/12154835/197cfd9c-b47f-11e5-8703-eb5052059b7d.png)

but I still have the same issue when calling `page.save_screenshot` using the capybara-webkit driver:

![testing_checkboxes_png_ _lightdeskapp](https://cloud.githubusercontent.com/assets/954468/12154774/b00ff530-b47e-11e5-9f31-c68404b4d348.png)


I'm hoping this is a QT issue?  Maybe I'm running a known-bad version? :)

**Comment by jferris** [comment]
There are generally known issues with Qt 4.x, but I haven't heard of this particular issue and I would expect Qt 5.4 to work correctly.

My versions are:

```
Capybara: 2.5.0
capybara-webkit: 1.7.1
Qt: 5.5.1
WebKit: 538.1
QtWebKit: 5.5.1
```

I'm on Linux, but I haven't encountered many Qt rendering bugs which are specific to a particular platform. I'm also on Qt 5.5 and Capybara 2.5.0. It seems very unlikely to be a Capybara bug. You could try updating Qt.

**Comment by jferris** [comment]
I'm going to close this, as it doesn't seem like a capybara-webkit bug, but I'm still watching the thread if you have further questions or ideas.

**Comment by noniq** [comment]
Seems like I'm encountering the exact same issue (all form elements positioned at the very top left edge).

I'm on Mac OS 10.10.5 and using:
```
Capybara: 2.5.0
capybara-webkit: 1.7.1
Qt: 5.5.1
WebKit: 538.1
QtWebKit: 5.5.1
```
Qt was installed and force-linked via Homebrew as described in the [Wiki](https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit).

Anyone else experiencing this?

**Comment by pmatsinopoulos** [comment]
I am experiencing this too. I have Mac OS X 10.10.5 using:

    Capybara: 2.6.2
    capybara-webkit: 1.8.0
    Qt: 5.5.0
    WebKit: 538.1
    QtWebKit: 5.5.0

Actually, I cannot see the checkboxes at all. For that reason, my tests fail. When I run selenium-webdriver everything seems to be ok.

**Comment by noniq** [comment]
For me all tests work correctly, even though the checkboxes are wrongly positioned (only using simple code like `check "mycheckbox"` and `#has_checked_field?("mycheckbox")`, though).

**Comment by jferris** [comment]
I'll try to do a local test soon using OS X 10.10.5. That seems to be the common element here.

**Comment by kingcheez** [comment]
I have the same issue on OSX 10.11. Your basic test case above shows up as below. I think it's a Qt problem:
![screen shot 2016-04-21 at 1 24 05 pm](https://cloud.githubusercontent.com/assets/279844/14718384/5dd87582-07c4-11e6-8abe-18b69ff67e83.png)

**Comment by cpina** [comment]
Hello - out of interest, I fixed the problem in our code base (not related to this code base but using webkit) doing: setStyle(QStyleFactory::create("Fusion")); to the WebKit widget. Just in case this helps...

**Comment by kingcheez** [comment]
I can confirm the same fix on OSX (also not related to this code base) :)

**Comment by jferris** [comment]
Thanks for the info, @cpina and @kingcheez!

Can some of the users in this thread who have experienced this issue try out the branch from #908?

**Comment by matt-riemer** [comment]
@jferris This totally fixes my issue!

I can see checkboxes and radio buttons again!  Thank you so much.

Confirmed working on:

```
Capybara: 2.7.0
capybara-webkit: 1.10.1
Qt: 5.5.1
WebKit: 538.1
QtWebKit: 5.5.1
```

**Comment by jferris** [comment]
This fix was released in 1.11.0.
